File: pascal/devices/pca9534.pas
{ PCA9534 I2C GPIO Expander device services                                   }

{ Copyright (C)2018, Philip Munts, President, Munts AM Corp.                  }
{                                                                             }
{ Redistribution and use in source and binary forms, with or without          }
{ modification, are permitted provided that the following conditions are met: }
{                                                                             }
{ * Redistributions of source code must retain the above copyright notice,    }
{   this list of conditions and the following disclaimer.                     }
{                                                                             }
{ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" }
{ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   }
{ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  }
{ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE   }
{ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR         }
{ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF        }
{ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS    }
{ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN     }
{ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)     }
{ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE  }
{ POSSIBILITY OF SUCH DAMAGE.                                                 }

UNIT PCA9534;

INTERFACE

  USES
    I2C,
    SysUtils;

  TYPE
    Error = CLASS(Exception);

    RegisterAddress = 0 .. 3;

    Device = CLASS
      CONSTRUCTOR Create(bus : I2C.Bus; addr : I2C.Address);

      { Read from a PCA9534 register }

      FUNCTION ReadRegister
       (reg  : RegisterAddress) : Byte;

      { Write to a PCA9534 register }

      PROCEDURE WriteRegister
       (reg  : RegisterAddress;
        data : Byte);

      { Read from PCA9534 input register }

      FUNCTION Read : Byte;

      { Write to PCA9534 output register }

      PROCEDURE Write
       (data : Byte);

      { Fetch last written value }

      FUNCTION Latch : Byte;

    PRIVATE
      mybus   : I2C.Bus;
      myaddr  : I2C.Address;
      mylatch : Byte;
    END;

  CONST
    InputPortReg     : RegisterAddress = 0;
    OutputPortReg    : RegisterAddress = 1;
    PolarityReg      : RegisterAddress = 2;
    ConfigurationReg : RegisterAddress = 3;

    AllInputs        : Byte = $FF;
    AllOutputs       : Byte = $00;
    AllNormal        : Byte = $00;  { Normal input polarity }
    AllOff           : Byte = $00;  { All outputs low }

IMPLEMENTATION

  CONSTRUCTOR Device.Create(bus : I2C.Bus; addr : I2C.Address);

  BEGIN
    Self.mybus  := bus;
    Self.myaddr := addr;
    Self.Write($FF);
  END;

  { Read from a PCA9534 register }

  FUNCTION Device.ReadRegister
   (reg  : RegisterAddress) : Byte;

  VAR
    cmd  : ARRAY [0 .. 0] OF Byte;
    resp : ARRAY [0 .. 0] OF Byte;

  BEGIN
    cmd[0] := reg;
    Self.mybus.Transaction(Self.myaddr, cmd, Length(cmd), resp, Length(resp));
    ReadRegister := resp[0];
  END;

  { Write to a PCA9534 register }

  PROCEDURE Device.WriteRegister
   (reg  : RegisterAddress;
    data : Byte);

  VAR
    cmd : ARRAY [0 .. 1] OF Byte;

  BEGIN
    IF reg = InputPortReg THEN
      RAISE Error.Create('ERROR: Cannot write to input register');

    cmd[0] := reg;
    cmd[1] := data;
    Self.mybus.Write(Self.myaddr, cmd, Length(cmd));

    IF reg = OutputPortReg THEN
      Self.mylatch := data;
  END;

  { Read from PCA9534 input register }

  FUNCTION Device.Read : Byte;

  BEGIN
    Read := Self.ReadRegister(InputPortReg);
  END;

  { Write to PCA9534 output register }

  PROCEDURE Device.Write(data : Byte);

  BEGIN
    Self.WriteRegister(OutputPortReg, data);
  END;

  { Fetch last written value }

  FUNCTION Device.Latch : Byte;

  BEGIN
    Latch := Self.mylatch;
  END;

END.

File: csharp/devices/pca9534.cs
// Copyright (C)2018-2023, Philip Munts dba Munts Technologies.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

namespace IO.Devices.PCA9534
{
    /// <summary>
    /// Encapsulates PCA9534 (and similar) I<sup>2</sup>C GPIO Expanders.
    /// </summary>
    public class Device
    {
        private readonly IO.Interfaces.I2C.Device dev;
        private byte config;
        private byte latch;
        private byte[] cmd = { 0, 0 };  // I2C command buffer
        private byte[] resp = { 0 };    // I2C response buffer

        /// <summary>
        /// The number of available GPIO pins per chip.
        /// </summary>
        public const int MAX_PINS = 8;

        /// <summary>
        /// Input Port Register address.
        /// </summary>
        public const byte InputPortReg = 0;

        /// <summary>
        /// Output Port Register address.
        /// </summary>
        public const byte OutputPortReg = 1;

        /// <summary>
        /// Input Port Polarity Register address.
        /// </summary>
        public const byte InputPolarityReg = 2;

        /// <summary>
        /// Configuration Register address.
        /// </summary>
        public const byte ConfigurationReg = 3;

        /// <summary>
        /// Configure all pins as inputs.
        /// </summary>
        public const byte AllInputs = 0xFF;

        /// <summary>
        /// Configure all pins as outputs.
        /// </summary>
        public const byte AllOutputs = 0x00;

        /// <summary>
        /// Configure all inputs as normal polarity.
        /// </summary>
        public const byte AllNormal = 0x00;

        /// <summary>
        /// Turn all outputs off.
        /// </summary>
        public const byte AllOff = 0x00;

        /// <summary>
        /// Constructor for a PCA9534 (or similar) GPIO Expander.
        /// </summary>
        /// <param name="bus">I<sup>2</sup>C bus controller.</param>
        /// <param name="addr">I<sup>2</sup>C slave address.</param>
        /// <param name="config">GPIO pin configuration.</param>
        /// <param name="states">Initial output states.</param>
        public Device(IO.Interfaces.I2C.Bus bus, int addr,
          byte config = AllInputs, byte states = AllOff)
        {
            this.dev = new IO.Interfaces.I2C.Device(bus, addr);
            this.Write(ConfigurationReg, config);
            this.Write(InputPolarityReg, AllNormal);
            this.Write(OutputPortReg, states);
        }

        /// <summary>
        /// Read from the specified PCA9534 device register.
        /// </summary>
        /// <param name="addr">Register address.</param>
        /// <returns>Register contents.</returns>
        public byte Read(byte addr)
        {
            if (addr > ConfigurationReg)
                throw new System.Exception("Invalid register address");

            this.cmd[0] = addr;
            this.dev.Transaction(this.cmd, 1, this.resp, 1);
            return this.resp[0];
        }

        /// <summary>
        /// Write to the specified PCA9534 device register.
        /// </summary>
        /// <param name="addr">Register address.</param>
        /// <param name="data">Data to written.</param>
        public void Write(byte addr, byte data)
        {
            if (addr > ConfigurationReg)
                throw new System.Exception("Invalid register address");

            if (addr == InputPortReg)
                throw new System.Exception("Cannot write to input port register");

            this.cmd[0] = addr;
            this.cmd[1] = data;
            this.dev.Write(cmd, 2);

            if (addr == ConfigurationReg)
                this.config = data;
            else if (addr == OutputPortReg)
                this.latch = data;
        }

        /// <summary>
        /// Return actual state of the GPIO pins.
        /// </summary>
        /// <returns>Pin states (MSB = GPIO7).</returns>
        public byte Read()
        {
            return this.Read(InputPortReg);
        }

        /// <summary>
        /// Write all GPIO pins.
        /// </summary>
        /// <param name="data">Data to write to pins (MSB = GPIO7).</param>
        public void Write(byte data)
        {
            this.Write(OutputPortReg, data);
        }

        /// <summary>
        /// This read-only property returns the last value written to the
        /// configuration register.
        /// </summary>
        public byte Config
        {
            get
            {
                return this.config;
            }
        }

        /// <summary>
        /// This read-only property returns the last value written to the
        /// output port register.
        /// </summary>
        public byte Latch
        {
            get
            {
                return this.latch;
            }
        }
    }
}
